### PR TITLE
Implicit targeting logic

### PR DESCRIPTION
**Pull request by charypar**
This adds basic capabilities for implicit targeting based on the request attributes.

## To do 

- [x] basic predicates
- [x] logic operators (AND, OR, NOT)
- [x] Regex matching
- [x] Accept and Accept-Language
- [x] JSON parsing and value extraction
- [x] List logic (in, all in, any in)
- [x] percentage match predicate ("N% of users identified by unique values of an attribute)
- [x] base64 deserialisation
- [x] browser and OS extraction
- [x] JSON (de)serialisation

### CODE REVIEW

**Comment by charypar** [comment]
As I was building this, I change my mind about how it should work in favour of more generic expressions encoded as an AST. 

These let us:
* Perform logic across multiple attributes if we want to
* Target on encoded attributes (e.g. `Accept` header or a JWT token)
* Target on calculated attributes (e.g. browser and operating system based on `User-Agent`)
* Combine expressions to avoid unnecessary additional built-ins
* Hopefully serialise to a more sensible configuration format as well

File: data-plane/src/features/implicit.rs [hunk 1]
```diff
@@ -199,10 +199,10 @@ impl StringExpr {
             } => todo!(),
             StringExpr::First(list) => list
                 .eval(request)
-                .and_then(|v| v.first().cloned().ok_or_else(|| "List is empty.".into())),
+                .and_then(|v| v.first().cloned().ok_or("List is empty.".into())),
```
Comment (StuartHarris) [inline data-plane/src/features/implicit.rs]: I added these in. It's more idiomatic and a clippy warning. Without the closure, the expression is executed regardless of whether it is needed. (https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call)

File: data-plane/src/features/implicit.rs [hunk 2]
```diff
@@ -279,14 +294,26 @@ mod test {
...
+    #[test_case(BoolExpr::Or(vec![BoolExpr::Constant(false), BoolExpr::Constant(false)]), Ok(false))]
     fn evaluate_boolean_expressions(expr: BoolExpr, expected: Result<bool, String>) {
         let request = [("hello", "world")].iter().cloned().collect();
 
         assert_eq!(expr.eval(&request), expected)
     }
 
     #[test]
-    #[ignore]
     fn matches_a_complex_expression() {
```
Comment (StuartHarris) [inline data-plane/src/features/implicit.rs:528]: This is amazing!

File: data-plane/src/features/implicit.rs (hunk 1 above)
Comment (charypar) [reply data-plane/src/features/implicit.rs]: Fair enough. Don't love how that reads, but take the point 😄

**Comment by charypar** [comment]
Right, this is hopefully all the functionality we might need for the moment. I have a feeling it would be worth doing some refactoring before merging this 😅

File: data-plane/src/features/implicit.rs [hunk 3]
```diff
@@ -86,13 +87,16 @@ pub enum BoolExpr {
...
-    fn eval(&self, request: &HashMap<&str, &str>) -> Result<bool, String> {
+    fn eval(&self, request: &HashMap<&str, &str>) -> Result<bool> {
         match &self {
             BoolExpr::Constant(c) => Ok(*c),
-            BoolExpr::Attribute(name) => request.get::<str>(name.as_ref()).map_or_else(
-                || Err(format!("Attribute '{}' not found.", name)),
-                |_| Ok(true),
-            ),
+            BoolExpr::Attribute(name) => {
+                if request.get::<str>(name.as_ref()).is_some() {
```
Comment (charypar) [inline data-plane/src/features/implicit.rs]: Can this be done with `ok_or_else`?

File: data-plane/src/features/implicit.rs [hunk 4]
```diff
@@ -343,20 +365,22 @@ fn parse_q_value(value: String) -> Vec<String> {
...
-                )
-            })
-        })
+    let value: serde_json::Value = serde_json::from_str(json)?;
+    if let Some(value) = value.pointer(pointer) {
+        if let Some(v) = cast(value) {
+            return Ok(v);
+        }
+    }
+    Err(anyhow::anyhow!(
```
Comment (charypar) [inline data-plane/src/features/implicit.rs]: Isn't the `anyhow!` macro itself in scope already?

**Review by charypar — COMMENTED** [review]
Really nice! 👍

**Pull request merged**